Replacing `get_utf8_value` method with wrappers to `six.ensure_` methods

walkerjoe opened the pull request:
Wrapper methods will attempt a list of common encodings, combined with
the appropriate `six.ensure_` method. Wrapper will also attempt to
handle non-string types--such as int and None.

Commit:
Added print_to_fd to utils.py

Added function print_to_fd and write_to_fd as a way to write
strings to file descriptors in a type agnostic way for
compatibility across Python versions.

Commit:
Added import collections

Commit:
Removed accidental paste, added sys import

Commit:
Spacing fix, comment fix

catleeball committed:
import io

walkerjoe committed:
Replacing `get_utf8_value` method with wrappers to `six.ensure_` methods

Wrapper methods will attempt a list of common encodings, combined with
the appropriate `six.ensure_` method. Wrapper will also attempt to
handle non-string types--such as int and None.

catleeball commented on boto/utils.py:
```diff
@@ -859,17 +862,35 @@ def notify(subject, body=None, html_body=None, to_string=None,
 
-    if isinstance(value, six.text_type):
-        value = value.encode('utf-8')
+    def _try_enc(string, encoding):
```
Two nits: This method might benefit from a docstring explaining what it does, and maybe consider a name for `conv_method` that describes what kind of function it should be handling.

walkerjoe replied:
Good point! Docstrings added.

catleeball commented:
Per our IM chat, it might be good to use chardet package instead.

walkerjoe committed:
Revised _TryEncodings method

Added docstrings and revised the logic a little to raise an error if
all encodings fail.
